### ISSUE TITLE
snippet support for mini.completion

### ISSUE THREAD

**Issue by alekspickle** [echasnovski/mini.nvim#886]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

completion

### Description

I already created the issue #882 about having $0 after completion.
I also then created an issue on LSP side where they [explained](https://github.com/mrcjkb/rustaceanvim/issues/393) that this is what is coming to Neovim 0.10 natively.

But mini.completion isn't supporting snippets, correct? if so, would you be open to support it or accept PR? currently I have no idea about volume of work to be done for it, but sound interesting and might help me get familiar with all this Lua madness.

**Comment by echasnovski** [comment]
Thanks for the suggestion!

Yes, [this is planned](https://github.com/echasnovski/mini.nvim/issues/574#issuecomment-1819531525) after the release of 'mini.snippets'. As 0.10 is released, this will now be possible to start working on. Yet, no concrete plans.

I appreciated the suggestion, but 'mini.completion' is one of the messiest module in 'mini.nvim' and I'll tackle it myself.

Let's have this issue open so as an extra measure to not forget about it.

**Comment by flexagoon** [comment]
@echasnovski can the 0.11 LSP completion side effects api be used for this?

**Comment by echasnovski** [comment]
> @echasnovski can the 0.11 LSP completion side effects api be used for this?

I am afraid I am not really sure what you're referring to. Could you link to it, please?

But if it is only 0.11, then it probably won't be used as I'd like 'mini.nvim' (and snippet support in 'mini.completion') to support as many versions as possible. Right now it should be possible to support 0.10+ but maybe after 'mini.snippets' this can get better.

**Comment by echasnovski** [comment]
> I was referring to the `vim.lsp.util.apply_text_edits` function. It seems like unless you need to have custom snippet support then using it together with the new snippets API should provide a very simple way to expand most existing LSP completions?

I *think* just using `vim.snippet` should be enough. It is just that resolving this issue is intentionally postponed after 'mini.snippets' is a thing. So that I'd have more knowledge about snippets and can be sure that I've tried to support as many previous Neovim versions as possible.

**Comment by flexagoon** [comment]
@echasnovski another thing that I thought:

A lot of the functionality of mini.completion is duplicating the functionality added in core with `vim.lsp.completion`. The main difference is that mini.completion has proper auto complete, while built-in is only activated on trigger chars.

So perhaps, when 0.11 is released, it would make sense to refactor that plugin to remove all of the custom popup menu item building and rely on `vim.lsp.completion.trigger()` instead? This would simplify the code, fit the plugin's goal of using as much native stuff as possible, and, as  an additional benefit, bring support for snippets and stuff for free.

**Comment by echasnovski** [comment]
> A lot of the functionality of mini.completion is duplicating the functionality added in core with `vim.lsp.completion`. The main difference is that mini.completion has proper auto complete, while built-in is only activated on trigger chars.

Purely due to how time works, I'd like to think that it is `vim.lsp.completion` which is duplicating 'mini.completion' functionality :) Although it used 'mfussenegger/nvim-lsp-compl' as a reference :) Or a more neutral "they share similar functionality" is even better.

> So perhaps, when 0.11 is released, it would make sense to refactor that plugin to remove all of the custom popup menu item building and rely on `vim.lsp.completion.trigger()` instead? This would simplify the code, fit the plugin's goal of using as much native stuff as possible, and, as an additional benefit, bring support for snippets and stuff for free.

Although motivation is sound, there are several crucial problems:
- 'mini.nvim' does not only support latest stable, but also two (minor) versions prior. So it will be not until the release of Neovim 0.13 when 'mini.completion' can be rewritten to use `vim.lsp.completion`.
- One *big* thing that I'd taken into account was that completion candidates persist across deleting characters. I.e. after completion candidates are shown, pressing `<BS>` should still show them with possibly bigger set of matched candidates. This is not the case with `vim.lsp.completion` (as far as my experiments and looking at its code showed) and I am not sure if it will be. Supporting this use case is one of the main reasons why 'mini.completion' uses 'omnifunc' and 'completefunc' approach instead of relying on `vim.fn.complete()` (which is what `vim.lsp.completion` currently uses).

**Comment by flexagoon** [comment]
> Purely due to how time works, I'd like to think that it is vim.lsp.completion which is duplicating 'mini.completion' functionality :) Although it used 'mfussenegger/nvim-lsp-compl' as a reference :) Or a more neutral "they share similar functionality" is even better.

I didn't mean that in the sense that you stole the functionality or anything like that :) I purely meant that by installing mini.completion you're now introducing extra code logic duplication. In the same sense I'd say that mini.comment is duplicating the functionality of built-in commenting, despite the fact that the built-in feature was implemented by you and based on mini.comment.

> One big thing that I'd taken into account was that completion candidates persist across deleting characters

I see, that makes sense. Perhaps when 0.11 is released you should add this note to the "difference from alternatives" section in the mini.completion docs?

---

The reason I asked is that I was playing around with native LSP completion and realized that the only things it lacks for me are pretty colors and proper autocompletion. With the ability to set pretty colors now being implemented (https://github.com/neovim/neovim/pull/30060), I was wondering if there could be a plugin that simply adds autocomplete logic on top of native completion.

But I guess just adding an autocmd with a few checks and a timer is enough for basic autocomplete, and it sounds simple enough that I could just do it in my nvim config without a plugin.

**Comment by echasnovski** [comment]
> ... purely meant that by installing mini.completion you're now introducing extra code logic duplication.

The overlap is not that big, I'd say, and it was already present from 'mini.completion' start (as its core initially was similar to `vim.lsp.omnifunc`). The "autotrigger" parts are quite different.

> In the same sense I'd say that mini.comment is duplicating the functionality of built-in commenting, despite the fact that the built-in feature was implemented by you and based on mini.comment.

Yeah, unfortunately, no good way to reuse built-in commenting because there is no exported Lua functions by design (otherwise no merge would happen). 'mini.comment' still offers more on top, though.

> The reason I asked is that I was playing around with native LSP completion and realized that the only things it lacks for me are pretty colors and proper autocompletion.

Fallback to built-in "usual" completion is a big part of 'mini.completion' and was one of the goals of even creating it. Because not every buffer has LSP attached to it yet autocompletion is still useful.

> With the ability to set pretty colors now being implemented ([neovim/neovim#30060](https://github.com/neovim/neovim/pull/30060)), I was wondering if there could be a plugin that simply adds autocomplete logic on top of native completion.

This doesn't look like a relevant PR. As far as I know, there is no way of setting highlight group in popup menu yet. This functionality needs to go through Vim and vim/vim#15561 seems to be the first step.

**Comment by flexagoon** [comment]
> This doesn't look like a relevant PR. As far as I know, there is no way of setting highlight group in popup menu yet. 

Completion items have a `hl_group` attribute, and this PR allows to set it

**Comment by echasnovski** [comment]
> Completion items have a `hl_group` attribute, and this PR allows to set it. The vim PR you linked references the issue I opened yesterday in nvim, and it introduces the option to set a hlgroup **just** for the kind text, but it's already possible to set a hlgroup for the entire completion item

Oh, indeed, my bad. It was added around a month ago, didn't catch that. This is a good time to incorporate 'mini.icons' into 'mini.completion' then. Thanks for the info!

**Comment by echasnovski** [comment]
@flexagoon, 'mini.completion' now by default highlights LSP kind (with `kind_hlgroup`) on Neovim>=0.11 if 'mini.icons' is enabled. It uses highlight groups from its ["lsp" category](https://github.com/echasnovski/mini.nvim/blob/57e47cf7a2923684e7413989ab267ed9730e7d03/doc/mini-icons.txt#L411). This requires enabling 'mini.icons', but I think it is a good alternative to possibly allowing customizing that highlighting in `MiniCompletion.config`. And I also don't really like the idea of something like `convert` from the `vim.lsp.completion.BufferOpts` as there is already a `MiniCompletion.config.lsp_completion.process_items` (which doesn't allow overriding `kind_hlgroup`, but I think it can be added if there is a demand).

**Comment by echasnovski** [comment]
> And I also don't really like the idea of something like `convert` from the `vim.lsp.completion.BufferOpts` as there is already a `MiniCompletion.config.lsp_completion.process_items` (which doesn't allow overriding `kind_hlgroup`, but I think it can be added if there is a demand).

After sleeping on it, allowing `config.lsp_completion.process_items` to set `kind_hlgroup` **is** the cleaner solution. The default 'mini.icons' one is shuffled to `MiniCompletion.default_process_items()` and still works as default (if 'mini.icons' is enabled). This should now be the part of `main`.

**Comment by GitMurf** [comment]
Just checking in to see if mini.snippets development is underway or any updates on roadmap for it :) thanks as usual!!

**Comment by echasnovski** [comment]
Yes, it is in development.

**Comment by echasnovski** [comment]
> I make snippets (`friendly_snippets` and `LSP snippets` tested) work inside mini.completion using none-ls which supports luasnip as a lsp source ...

Something like is planned for 'mini.completion' and 'mini.snippets', but not during initial 'mini.snippets' release. But hopefully more or less shortly after.

**Comment by echasnovski** [comment]
> could you guide me, what argument should go in `vim.snippet.expand()` to use it instead of `feedkeys '<c-y>'` in https://github.com/YeferYV/RetroNvim/blob/50b6715793cdceb77ba7a665a671a52fd5e8c566/nvim/init.lua#L396C1-L396C25 (which is based in https://gist.github.com/MariaSolOs/2e44a86f569323c478e5a078d0cf98cc) ? I want to see if `vim.snippet.expand()` could expand `friendly-snippets`'s snippets

Sorry, I don't use built-in Neovim>=0.11 LSP completion, so can't really help here.

**Comment by echasnovski** [comment]
Just an FYI: ['mini.snippets' is now a thing](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-snippets.md). I plan to tackle this (and other 'mini.completion' issues) as the next "big" project. Let's hope it will not end up actually big, though :)

**Comment by masroof-maindak** [comment]
Hello, just wondering if there's any update on this front (i.e integrating mini.snippets w/ mini.completion). Many thanks for your time and effort.

**Comment by echasnovski** [comment]
> Hello, just wondering if there's any update on this front (i.e integrating mini.snippets w/ mini.completion). Many thanks for your time and effort.

Yes, there is. I am building up courage to fully understand LSP spec regarding snippets so that I wouldn't have to deal with it soon-ish again. And in the meantime making 'mini.completion' more robust for these changes, along with other interesting updates.

**Comment by echasnovski** [comment]
This is (**finally**) resolved! Main attention has gone to supporting 'mini.snippets' as "snippet insert" backend, but there is `vim.snippet` fallback by default. Here is the [relevant part of help](https://github.com/echasnovski/mini.nvim/blob/6dcb014319a177ad51774c05323cb833349bdf25/doc/mini-completion.txt#L80-L105) with pointers for where to look next. [The demo](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-completion.md#demo) was also updated to better reflect current 'mini.completion' capabilities.

Thanks to all of you for waiting patiently here 🙏 and sorry it took so long.

**Comment by ldelossa** [comment]
awesome!!! I've been waiting for this. I'm trying it out now. Took me a second to realize you enter a 'session' during snippet insert, which I think is a really cool way to do this. 

One question I have. I currently have a super-tab thing going on with Vim's built in snippets:
```lua
local keys = {
	["tab"] = vim.api.nvim_replace_termcodes("<Tab>", true, true, true),
	["ctrl-y"] = vim.api.nvim_replace_termcodes("<C-y>", true, true, true),
	["noop"] = vim.api.nvim_replace_termcodes("", true, true, true),
}
vim.g.super_tab = function()
	if vim.fn.pumvisible() ~= 0 then
		return keys["ctrl-y"]
	elseif require("copilot.suggestion").is_visible() then
		require("copilot.suggestion").accept()
		return keys["noop"]
	elseif vim.snippet.active({ direction = 1 }) then
		return '<cmd>lua vim.snippet.jump(1)<cr>'
	else
		return keys["tab"]
	end
end
vim.keymap.set("i", "<Tab>", vim.g.super_tab, { expr = true })
```

Does MiniSnippets provide an API where I can see if I'm in an active MiniSnippet session and call a function to jump to next tabstop?

**Comment by echasnovski** [comment]
> Does MiniSnippets provide an API where I can see if I'm in an active MiniSnippet session and call a function to jump to next tabstop?

Yep. Try replacing ...

```lua
	elseif vim.snippet.active({ direction = 1 }) then
		return '<cmd>lua vim.snippet.jump(1)<cr>'
```

with this:

```lua
	elseif MiniSnippets.session.get() ~= nil then
		return '<cmd>lua MiniSnippets.session.jump("next")<cr>'
```

Here is the [relevant part of 'mini.snippets' help](https://github.com/echasnovski/mini.nvim/blob/6dcb014319a177ad51774c05323cb833349bdf25/doc/mini-snippets.txt#L414).

**Comment by ldelossa** [comment]
Thank you @echasnovski I totally missed that 🤦 

I'm playing with the new snippets + completion and its lovely :)

**Comment by wenjinnn** [comment]
Thank you for another great work @echasnovski !
After some test, I find out completion snippet did not work with the java lsp `jdtls`, it returns completion item data like this:
```lua
{                                                                                                                                                                                                         
  command = {                                                                                                                                                                                             
    arguments = { "11", "0" },                                                                                                                                                                            
    command = "java.completion.onDidSelect",                                                                                                                                                              
    title = ""                                                                                                                                                                                            
  },                                                                                                                                                                                                      
  data = {                                                                                                                                                                                                
    pid = "0",                                                                                                                                                                                            
    rid = "11"                                                                                                                                                                                            
  },                                                                                                                                                                                                      
  detail = "String.valueOf(double d) : String",                                                                                                                                                           
  filterText = "valueOf(${1:0})",                                                                                                                                                                         
  insertText = "valueOf",                                                                                                                                                                                 
  kind = 2,                                                                                                                                                                                               
  kind_hlgroup = "MiniIconsAzure",                                                                                                                                                                        
  label = "valueOf(double d) : String",                                                                                                                                                                   
  sortText = "999999179",                                                                                                                                                                                 
  textEditText = "valueOf(${1:0})"                                                                                                                                                                        
}
```
It only have snippet tabstops string in `textEditText` and `filterText` field, have no `InsertTextFormat`, in the meantime, the `label` field has extra function signature string, which it mess up completion item word. in this case, the word should be `insertText`

I've already fixup this in this [commit](https://github.com/wenjinnn/mini.nvim/commit/0475d83064893d7f5e80818248e966862f27e03c), but I'm not sure it's the best solution. If you think this is good, I'll be happy to make a PR.

**Comment by echasnovski** [comment]
> It only have snippet tabstops string in `textEditText` and `filterText` field, have no `InsertTextFormat`, in the meantime, the `label` field has extra function signature string, which it mess up completion item word. in this case, the word should be `insertText`

Hmmm... Then judging by [LSP specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem), this is not a snippet. A snippet is defined in exactly two ways: either by kind (should be 15; 2 is for "Method") or by `InsertTextFormat` (should be present and equal to 2). I don't think deviating from this specification is a good idea; especially by treating a snippet any item that has tabstop-like text in it.

Or is there an `insertTextFormat` set to 2 in the whole `CompletionList` response (one step higher with respect to nesting)?

Respecting the `textEditText` is a good idea, though. I'll take a look.

**Comment by masroof-maindak** [comment]
Hello, thank you for the release.

In my (admittedly short) test-drive so far, it's been very inconsistent in terms of behaviour.

- Works flawlessly in Lua
- In Python or Go, the snippets simply don't show up in my completion list.
- For C, the snippet does show up in my completion list, but selecting w/ `<C-y>` doesn't activate the 'session.'

For case 2 and 3, entering the snippet's shorthand manually and expanding it with `<C-j>` works fine.

Is this a config-issue or an LSP-issue? I don't believe it to be the former, but just in case, here are the relevant lines from my config.

- https://github.com/masroof-maindak/nvim/blob/c47c5d7b244065c3eb5237b5f96347041aebd228/lua/autocommands.lua#L35
- https://github.com/masroof-maindak/nvim/blob/c47c5d7b244065c3eb5237b5f96347041aebd228/lua/plugins/snippets.lua
- https://github.com/masroof-maindak/nvim/blob/c47c5d7b244065c3eb5237b5f96347041aebd228/lua/plugins/mini-utils.lua#L23-L34

**Comment by echasnovski** [comment]
> For case 2 and 3, entering the snippet's shorthand manually and expanding it with `<C-j>` works fine.

This leads to believe that you expect snippets loaded *by* 'mini.snippets' to show up among completion items. This is not the case (yet) as this issue was about respecting snippet items that were *already* in the completion list. See [this comment](https://github.com/echasnovski/mini.nvim/discussions/1647#discussioncomment-12427468) for more details.

> P.S, I've had to disable this keybind

That's the result of [added scrolling](https://github.com/echasnovski/mini.nvim/commit/621d7bb83159e2d3fec804cc77878f151bc1d5b5). I'd suggest that the cleaner solution is to disable [scrolling mappings](https://github.com/echasnovski/mini.nvim/blob/f6fe169821506a59b2407e3c7a3327562600e0e3/doc/mini-completion.txt#L262-L263) or use different keys. I forgot that it might conflict with user mappings, so didn't mark commit as breaking. Sorry.

**Comment by masroof-maindak** [comment]
> This leads to believe that you expect snippets loaded _by_ 'mini.snippets' to show up among completion items. This is not the case (yet) as this issue was about respecting snippet items that were _already_ in the completion list. See [this comment](https://github.com/echasnovski/mini.nvim/discussions/1647#discussioncomment-12427468) for more details.

Ah, I see, thanks for the clarification. Out of curiosity, how come some of the snippets, as you said, happen to *already* exist in the completion list? Do their LSPs support snippets OOTB or something?

**Comment by echasnovski** [comment]
> > This leads to believe that you expect snippets loaded _by_ 'mini.snippets' to show up among completion items. This is not the case (yet) as this issue was about respecting snippet items that were _already_ in the completion list. See [this comment](https://github.com/echasnovski/mini.nvim/discussions/1647#discussioncomment-12427468) for more details.
> 
> Ah, I see, thanks for the clarification. Out of curiosity, how come some of the snippets, as you said, happen to _already_ exist in the completion list? Do their LSPs support snippets OOTB or something?

Yes, some LSP servers on their own provide snippets (as in - items with "Snippet" kind) as completion suggestions. Like indeed LuaLS (`do .. end`, `function ()`, etc.).

**Comment by wenjinnn** [comment]
> Hmmm... Then judging by [LSP specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem), this is not a snippet. A snippet is defined in exactly two ways: either by kind (should be 15; 2 is for "Method") or by `InsertTextFormat` (should be present and equal to 2). I don't think deviating from this specification is a good idea; especially by treating a snippet any item that has tabstop-like text in it.
>
> 
> Or is there an `insertTextFormat` set to 2 in the whole `CompletionList` response (one step higher with respect to nesting)?

I agree. After debugging the `CompletionList`, I found that they do have an `itemDefaults` which includes `insertTextFormat = 2`:
```lua
{
  isIncomplete = false,
  itemDefaults = {
    data = {
      completionKinds = { 2, 3, 6, 9 }
    },
    editRange = {
      ["end"] = {
        character = 16,
        line = 79
      },
      start = {
        character = 15,
        line = 79
      }
    },
    insertTextFormat = 2
  },
  items = { {
      command = {
        arguments = { "13", "0" },
        command = "java.completion.onDidSelect",
        title = ""
      },
      data = {
        pid = "0",
        rid = "13"
      },
      detail = "String.valueOf(double d) : String",
      filterText = "valueOf(${1:0})",
      insertText = "valueOf",
      kind = 2,
      label = "valueOf(double d) : String",
      sortText = "999999179",
      textEditText = "valueOf(${1:0})"
    }
  }
}
```
 Then I lookup into Neovim repo, and in they built-in completion, they handle it like [this](https://github.com/neovim/neovim/blob/c38c88edfd015e8badbd11941030d7d85b5b1dc7/runtime/lua/vim/lsp/completion.lua#L191-L230). Shamelessly, I copied the relevant code over, and now it should close to a reasonable [solution](https://github.com/wenjinnn/mini.nvim/commit/e1089e481ed62f7fbcf2ce3c9e42889ba4cb1760) :)

**Comment by echasnovski** [comment]
> I agree. After debugging the `CompletionList`, I found that they do have an `itemDefaults` which includes `insertTextFormat = 2`:
> ...
> Then I lookup into Neovim repo, and in they built-in completion, they handle it like [this](https://github.com/neovim/neovim/blob/c38c88edfd015e8badbd11941030d7d85b5b1dc7/runtime/lua/vim/lsp/completion.lua#L191-L230). Shamelessly, I copied the relevant code over, and now it should close to a reasonable [solution](https://github.com/wenjinnn/mini.nvim/commit/e1089e481ed62f7fbcf2ce3c9e42889ba4cb1760) :)

Yeah, I noticed that after you posted the comment. I'll add support for item defaults in a similar way soon-ish.

**Comment by echasnovski** [comment]
@wenjinnn, latest `main` should now have support for item defaults. I didn't test it with your case of `jdtls`, but it should work according to specification.

---

Just as a not, there was no update to use `insertText` as `word` field of completion item for snippets (as the is done in the implementation from [this comment](https://github.com/echasnovski/mini.nvim/issues/886#issuecomment-2707989215)). Mostly because I think this is a bit too cherry-picked to work better in this particular case: other LSP servers might (and probably do) have `insertText` as an actual snippet string which defeats the purpose there.

**Comment by wenjinnn** [comment]
> [@wenjinnn](https://github.com/wenjinnn), latest `main` should now have support for item defaults. I didn't test it with your case of `jdtls`, but it should work according to specification.

Thanks for the reminder! I've tested latest `main`, snippets now work with `jdtls`.

> Just as a not, there was no update to use `insertText` as `word` field of completion item for snippets (as the is done in the implementation from [this comment](https://github.com/echasnovski/mini.nvim/issues/886#issuecomment-2707989215)). Mostly because I think this is a bit too cherry-picked to work better in this particular case: other LSP servers might (and probably do) have `insertText` as an actual snippet string which defeats the purpose there.

Can I ask you to consider this again? because I can't find a way to fix this outside the mini.completion, and it might be the standard way to deal with `word`. As a [reference](https://github.com/neovim/neovim/blob/092529ebbf6fe566399aa47c274d708b8dea5334/runtime/lua/vim/lsp/completion.lua#L146-L189), neovim built-in completion also handles completion word using this priority order. The author also refer this [lsp specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion)

**Comment by echasnovski** [comment]
> Can I ask you to consider this again? because I can't find a way to fix this outside the mini.completion, and it might be the standard way to deal with `word`. As a [reference](https://github.com/neovim/neovim/blob/092529ebbf6fe566399aa47c274d708b8dea5334/runtime/lua/vim/lsp/completion.lua#L146-L189), neovim built-in completion also handles completion word using this priority order. The author also refer this [lsp specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion)

I don't think there is a need to fix anything here. [This line](https://github.com/wenjinnn/mini.nvim/blob/e1089e481ed62f7fbcf2ce3c9e42889ba4cb1760/lua/mini/completion.lua#L1006) is intended to only set `word` temporarily because `word` is a required field for Neovim's completion item. It defines a word which will be inserted during up-down navigation across completion items.

The actual snippet text is computed [later](https://github.com/echasnovski/mini.nvim/blob/d7ba1dec0082b7a542aa79c327e9554d4a27cc2c/lua/mini/completion.lua#L1077) (as there might be updates from `completionItem/resolve`) and `insertText` [*is* used there to find the actual snippet text](https://github.com/echasnovski/mini.nvim/blob/d7ba1dec0082b7a542aa79c327e9554d4a27cc2c/lua/mini/completion.lua#L1068). Preferring text from `textEdit` over `insertText` [is according to spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem) ("... Therefore it is recommended to use `textEdit` instead since it avoids additional client side interpretation." as a part of `insertText` description).

So for [this server response](https://github.com/echasnovski/mini.nvim/issues/886#issuecomment-2707989215) the `valueOf(double d) : String` (i.e. `label`) will be used in popup menu and when navigating up-down completion list while `valueOf(${1:0})` (i.e. `textEditText`) will be used as snippet text.

**Comment by wenjinnn** [comment]
Thanks for the detailed explanation.

> So for [this server response](https://github.com/echasnovski/mini.nvim/issues/886#issuecomment-2707989215) the `valueOf(double d) : String` (i.e. `label`) will be used in popup menu and when navigating up-down completion list while `valueOf(${1:0})` (i.e. `textEditText`) will be used as snippet text.

Yap, that's the normal workflow, which is also work great for me, but my case is once I didn't apply current selected item but choose to return to normal mode, the selected text will leave there, Unless I have `noinsert` in `completeopt`.

![Image](https://github.com/user-attachments/assets/94e3ba2d-87a6-4c43-a691-6041055a115d)

**Comment by echasnovski** [comment]
> ... but my case is once I didn't apply current selected item but choose to return to normal mode, the selected text will leave there, Unless I have `noinsert` in `completeopt`.

This is an intended behavior with the following reasoning:
- There is no item selected when popup is shown (as 'noselect' is set in 'completeopt' in `require('mini.completion').setup()`). So selecting an item is treated as an intended action from the user.
- Selecting an item with snippet text is temporary because snippet should be actually inserted, which ends a completion. Inserting is a separate action done after `<C-y>` or non-keyword character. There is no good reason to support inserting "temporary" text as permanent for snippet text items because LSP server said that that item should be inserted as snippet.
- If you want remove inserted text and end a completion (even before exiting into Normal mode), [press `<C-e>` for that](https://github.com/echasnovski/mini.nvim/blob/d7ba1dec0082b7a542aa79c327e9554d4a27cc2c/doc/mini-completion.txt#L94-L95).

**Comment by wenjinnn** [comment]
> > ... but my case is once I didn't apply current selected item but choose to return to normal mode, the selected text will leave there, Unless I have `noinsert` in `completeopt`.
> 
> This is an intended behavior with the following reasoning:
> 
>     * There is no item selected when popup is shown (as 'noselect' is set in 'completeopt' in `require('mini.completion').setup()`). So selecting an item is treated as an intended action from the user.
> 
>     * Selecting an item with snippet text is temporary because snippet should be actually inserted, which ends a completion. Inserting is a separate action done after `<C-y>` or non-keyword character. There is no good reason to support inserting "temporary" text as permanent for snippet text items because LSP server said that that item should be inserted as snippet.
> 
>     * If you want remove inserted text and end a completion (even before exiting into Normal mode), [press `<C-e>` for that](https://github.com/echasnovski/mini.nvim/blob/d7ba1dec0082b7a542aa79c327e9554d4a27cc2c/doc/mini-completion.txt#L94-L95).

Ok, I'm just too lazy so that I don't want to press another `<C-e>` before returning to normal mode. but you convinced me. I'll just using `<C-e>` more than before :)

**Comment by olekspickle** [comment]
Finally got around to migrate to mini.snippets. So far it's a blast!
So much more convenient than convoluted luasnip syntax.

**Comment by echasnovski** [comment]
For anyone still subscribed to this issue, there is now (finally) a way to expose snippets collected/prepared/loaded by 'mini.snippets' to (auto)completion engines like 'mini.completion': add a [`MiniSnippets.start_lsp_server()`](https://github.com/echasnovski/mini.nvim/blob/6be765891c95242f6b8ebd3a25aa285ab3e9f6de/doc/mini-snippets.txt#L1240) call after `require('mini.snippets').setup()`. This will start a dedicated LSP server for 'mini.snippets' which will provide its collected snippets as completion suggestions.

**Comment by 231tr0n** [comment]
@echasnovski I apologize for the ping but is it possible to show the actual snippet along with the snippet info as part of the completion item doc window.

![Image](https://github.com/user-attachments/assets/b2c96e29-6999-47bb-a1d9-3a5fbec01161)

**Comment by echasnovski** [comment]
> [@echasnovski](https://github.com/echasnovski) I apologize for the ping but is it possible to show the actual snippet along with the snippet info as part of the completion item doc window.

Hmm... Yes, this is possible by providing snippet as a `detail`. It might get too big for some snippets out there, but I'll take a look.

**Comment by echasnovski** [comment]
> > [@echasnovski](https://github.com/echasnovski) I apologize for the ping but is it possible to show the actual snippet along with the snippet info as part of the completion item doc window.
> 
> Hmm... Yes, this is possible by providing snippet as a `detail`. It might get too big for some snippets out there, but I'll take a look.

The latest `main` should include snippet body as `detail` (in 'mini.completion' will be shown in info window on top of description, with syntax highlighting), but only if it is different from snippet's description (which can happen if there was no description in original snippet). Thanks for the suggestion!
